lib/bootstrap.rb

```diff
@@ -0,0 +1,118 @@
+module LogStash
+  module Environment
+    extend self
+
+    # also set the env LOGSTASH_HOME
+    LOGSTASH_HOME = ENV["LOGSTASH_HOME"] = ::File.expand_path(::File.join(__FILE__, "..", ".."))
+
+    BUNDLE_DIR = ::File.join(LOGSTASH_HOME, "vendor", "bundle")
+    GEMFILE_PATH = ::File.join(LOGSTASH_HOME, "Gemfile")
+
+    # @return [String] the ruby version string bundler uses to craft its gem path
+    def gem_ruby_version
+      RbConfig::CONFIG["ruby_version"]
+    end
+
+    # @return [String] major.minor ruby version, ex 1.9
+    def ruby_abi_version
+      RUBY_VERSION[/(\d+\.\d+)(\.\d+)*/, 1]
+    end
+
+    # @return [String] jruby, ruby, rbx, ...
+    def ruby_engine
+      RUBY_ENGINE
+    end
+
+    def logstash_gem_home
+      ::File.join(BUNDLE_DIR, ruby_engine, gem_ruby_version)
+    end
+  end
+
+  module Bootstrap
+    extend self
+
+    def bundler_setup!(options = {})
+      options = {:without => [:development]}.merge(options)
+      options[:without] = Array(options[:without])
+
+      # make sure we use our own installed bundler
+      patch_rubygem!
+      ::Gem.clear_paths
+      ::Gem.paths = ENV['GEM_HOME'] = ENV['GEM_PATH'] = Environment.logstash_gem_home
+
+      # set BUNDLE_GEMFILE ENV before requiring bundler to avoid bundler recurse and load unrelated Gemfile(s)
+      ENV["BUNDLE_GEMFILE"] = Environment::GEMFILE_PATH
+
+      require "bundler"
+      patch_bundler!
+
+      ::Bundler.settings[:path] = Environment::BUNDLE_DIR
+      ::Bundler.settings[:without] = options[:without].join(":")
+      # in the context of Bundler.setup it looks like this is useless here because Gemfile path can only be specified using
+      # the ENV, see https://github.com/bundler/bundler/blob/v1.8.3/lib/bundler/shared_helpers.rb#L103
+      ::Bundler.settings[:gemfile] = Environment::GEMFILE_PATH
+
+      ::Bundler.reset!
+      ::Bundler.setup
+    end
+
+
+    def patch_bundler!
+      # Patch bundler to write a .lock file specific to the version of ruby.
+      # This keeps MRI/JRuby/RBX from conflicting over the Gemfile.lock updates
+      ::Bundler::SharedHelpers.module_exec do
+        def default_lockfile
+          ruby = "#{Environment.ruby_engine}-#{Environment.ruby_abi_version}"
+          Pathname.new("#{default_gemfile}.#{ruby}.lock")
+        end
+      end
+    end
+
+    def patch_rubygem!
```

Comment:
why not having this code into the patches directory ?, there we've some others patches and helps to keep things easy to locate when have to go away.